File: test/matrix-test.ts
import { suite, test } from "mocha-typescript";
import { expect } from 'chai';
import { Vector2D } from "../src/vector";
import { Matrix3, Matrix4 } from "../src/matrix";

@suite
class MatrixTest {


    @test identity() {
        expect(Matrix3.identity()).to.deep.equal(
            new Matrix3(
                1,0,0,
                0,1,0,
                0,0,1
            )
        )
    }

    @test translation() {
        expect(Matrix3.identity().translate(new Vector2D(10,-5)).translate(new Vector2D(6,3))).to.deep.equal(
            new Matrix3(
                1,0,16,
                0,1,-2,
                0,0,1
            )
        )
    }

    @test scale() {
        expect(Matrix3.identity().scale(new Vector2D(10,-5)).scale(new Vector2D(2,0.1))).to.deep.equal(
            new Matrix3(
                20,0,0,
                0,-0.5,0,
                0,0,1
            )
        )
    }

    @test transform() {
        let transformationMatrix = Matrix3.identity().scale(new Vector2D(0.5,2)).translate(new Vector2D(-2,4))
        let vector = new Vector2D(5,3)
        expect(transformationMatrix.transform(vector)).to.deep.equal(new Vector2D(1.5,14))
    }

    @test transpose() {
        let M = new Matrix3(
            1,2,3,
            4,5,6,
            7,8,9
        )
        expect(M.transpose()).to.deep.equal(
            new Matrix3(
                1,4,7,
                2,5,8,
                3,6,9
            )
        )
    }

    @test determinant() {
        let M = new Matrix3(
            1,2,3,
            0,1,4,
            5,6,0
        )
        expect(M.determinant()).to.equal(1)
    }

    @test determinant4x4() {
        let M = new Matrix4(
            6,0,-3,5,
            4,13,6,-8,
            -1,0,7,4,
            8,6,0,2
        )
        expect(M.determinant()).to.equal(170)
    }

    @test adjugate() {
        let M = new Matrix3(
            1,2,3,
            0,1,4,
            5,6,0
        )
        expect(M.adjugate()).to.deep.equal(
            new Matrix3(
                -24, 18, 5,
                20,-15,-4,
                -5,4,1
            )
        )
    }

    @test inverse() {
        let M = new Matrix3(
            1,2,3,
            0,1,4,
            5,6,0
        )
        let MInverse = new Matrix3(
            -24, 18, 5,
            20,-15,-4,
            -5,4,1
        )
        expect(M.inverse()).to.deep.equal(MInverse)
        expect(MInverse.multiply(M)).to.deep.equal(Matrix3.identity())
    }

    @test inverse4x4() {
        let M = new Matrix4(
            1,1,1,0,
            0,3,1,2,
            2,3,1,0,
            1,0,2,1
        )
        let MInverse = new Matrix4(
            -3, -0.5, 1.5, 1,
            1, 0.25, -0.25, -0.5,
            3, 0.25, -1.25, -0.5,
            -3,-0,1,1
        )
        expect(M.inverse()).to.deep.equal(MInverse)
        expect(MInverse.multiply(M)).to.deep.equal(Matrix4.identity())
    }

}

File: test/MatrixTest.kt
import sgl.*
import kotlin.test.Test
import kotlin.test.expect

class MatrixTest {

    @Test fun identity3() = expect(Matrix3(
            1f, 0f, 0f,
            0f, 1f, 0f,
            0f, 0f, 1f
    )){ Matrix3.identity }

    @Test fun translation3() =  expect(Matrix3(
            1f, 0f, 16f,
            0f, 1f, -2f,
            0f, 0f, 1f
    )){ Matrix3.identity translate Vector2(10f, -5f) translate Vector2(6f, 3f) }

    @Test fun scale3() = expect(Matrix3(
            20f, 0f, 0f,
            0f, -0.5f, 0f,
            0f, 0f, 1f
    )){ Matrix3.identity scale Vector2(10f, -5f) scale Vector2(2f, 0.1f) }

    @Test fun transform3() = expect(Vector2(1.5f, 14f)) {
        (Matrix3.identity scale Vector2(0.5f, 2f) translate Vector2(-2f, 4f)) * Vector2(5f, 3f)
    }

    @Test fun transpose3() =  expect(Matrix3(
            1f, 4f, 7f,
            2f, 5f, 8f,
            3f, 6f, 9f
    )){
        Matrix3(
                1f, 2f, 3f,
                4f, 5f, 6f,
                7f, 8f, 9f
        ).transpose()
    }

    @Test fun determinant3() = expect(1f){
        Matrix3(
                1f, 2f, 3f,
                0f, 1f, 4f,
                5f, 6f, 0f
        ).determinant
    }

    @Test fun adjugate3() = expect(
            Matrix3(
                    -24f, 18f, 5f,
                    20f, -15f, -4f,
                    -5f, 4f, 1f
            )
    ){
        Matrix3(
                1f, 2f, 3f,
                0f, 1f, 4f,
                5f, 6f, 0f
        ).adjugate()
    }

    @Test fun inverse3(){
        val inverse = Matrix3(
                -24f, 18f, 5f,
                20f, -15f, -4f,
                -5f, 4f, 1f
        )
        val matrix = Matrix3(
                1f, 2f, 3f,
                0f, 1f, 4f,
                5f, 6f, 0f
        )
        expect(inverse){ matrix.inverse() }
        expect(Matrix3.identity){ matrix * inverse }
    }

    @Test fun identity4() = expect(Matrix4(
            1f, 0f, 0f, 0f,
            0f, 1f, 0f, 0f,
            0f, 0f, 1f, 0f,
            0f, 0f, 0f, 1f,
            identity = true
    )){ Matrix4.identity }

    @Test fun translation4() =  expect(Matrix4(
            1f, 0f, 0f, 16f,
            0f, 1f, 0f, -2f,
            0f, 0f, 1f, 15f,
            0f, 0f, 0f, 1f
    )){ Matrix4.identity translate Vector3(10f, -5f, 10f) translate Vector3(6f, 3f, 5f) }

    @Test fun scale4() = expect(Matrix4(
            20f, 0f, 0f, 0f,
            0f, -0.5f, 0f, 0f,
            0f, 0f, 10f, 0f,
            0f, 0f, 0f, 1f
    )){ Matrix4.identity scale Vector3(10f, -5f, -5f) scale Vector3(2f, 0.1f, -2f) }


    @Test fun transform4() = expect(Vector3(1.5f, 14f, 15f)) {
        (Matrix4.identity scale Vector3(0.5f, 2f, 3f) translate Vector3(-2f, 4f, -5f)) * Vector3(5f, 3f, 10f)
    }

    @Test fun transpose4() =  expect(Matrix4(
            1f, 5f, 9f, 13f,
            2f, 6f, 10f, 14f,
            3f, 7f, 11f, 15f,
            4f, 8f, 12f, 16f
    )){
        Matrix4(
                1f, 2f, 3f, 4f,
                5f, 6f, 7f, 8f,
                9f, 10f, 11f, 12f,
                13f, 14f, 15f, 16f
        ).transpose()
    }

    @Test fun determinant4() = expect(170f){
        Matrix4(
                6f, 0f, -3f, 5f,
                4f, 13f, 6f, -8f,
                -1f, 0f, 7f, 4f,
                8f, 6f, 0f, 2f
        ).determinant
    }

    @Test fun inverse4(){
        val inverse = Matrix4(
                -3f, -0.5f, 1.5f, 1f,
                1f, 0.25f, -0.25f, -0.5f,
                3f, 0.25f, -1.25f, -0.5f,
                -3f, -0f, 1f, 1f
        )
        val matrix = Matrix4(
                1f, 1f, 1f, 0f,
                0f, 3f, 1f, 2f,
                2f, 3f, 1f, 0f,
                1f, 0f, 2f, 1f
        )
        expect(inverse){ matrix.inverse() }
        expect(Matrix4.identity.copy(identity = false)){ matrix * inverse }
    }

}